Use font based on italic and font weight settings

rkusa opened the pull request:
This PR is not ready, yet, but I'd like to have some feedback before continuing it (if continuing at all).

I've started working on the following code comments

```
// TODO: apply actual numeric weight
// TODO: use true italic in font if available
```

... and implemented the changes while keeping the future required string width calculation in mind.

The main concern is maybe that I made the `Style` struct font-aware. While this does work well for style-spans, it does not work well with the current `TextDrawingMetrics` that lives in the `EditViewController`. My feeling is that the `TextDrawingMetrics` could be replaced with a default Style (maybe Style 0 as the default style, and 1 the selection).

rkusa commented on XiEditor/AppDelegate.swift:
```diff
@@ -74,7 +74,8 @@ class AppDelegate: NSObject, NSApplicationDelegate {
         case "def_style":
             if let obj = params as? [String : AnyObject] {
-                styleMap.defStyle(json: obj)
+                let font = CTFontCreateWithName("InconsolataGo" as CFString?, 14, nil)
```
This is temporary. But if we keep the Style objects font-aware, we might re-consider whether each buffer needs its own styleMap instance instead of using a global one.

raphlinus replied:
That is open to reconsideration, yes.

rkusa commented on XiEditor/StyleMap.swift:
```diff
@@ -16,11 +16,7 @@ import Cocoa
-    var weight: UInt16;
-    var underline: Bool;
-    var italic: Bool;
+    var attributes: [String: Any];
```
We could also keep all the properties, but instead of creating the attributes array on every call, it is now just created once.

rkusa commented on XiEditor/StyleMap.swift:
```diff
@@ -108,4 +122,25 @@ class StyleMap {
+        }
+        weight += direction
+    }
+    return nil
```
This one looks for the font that best matches the targeted weight settings. (I should have added a doc comment before the function).

cmyr replied:
you might be able to clean this up with [availableMembers(ofFontFamily:)](https://developer.apple.com/reference/appkit/nsfontmanager/1462316-availablemembers).

rkusa replied:
I did an implementation with `availableMembers` first, but ended up with a less understandable code. If you like, I can show of the `availableMembers` variant of this implementation? (I also did a `convertWeight` implementation,  but this API did some really strange things).

cmyr replied:
i'd be curious, if you have it saved? Just throw it in a gist somewhere? But no worries if you didn't stash it.

rkusa replied:
https://gist.github.com/rkusa/4266960dc20085074a90dad36fd6d2a5

A final version could use some cleanup and also some logic to keep original traits (like Condensed). But probably enough to get an idea.

cmyr commented on XiEditor/StyleMap.swift:
```diff
@@ -61,15 +57,46 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
-        let underline = json["underline"] as? Bool ?? false
+
+        var attrs: [String: Any] = [
+            NSForegroundColorAttributeName: json["fg_color"] as? NSColor ?? NSColor.red,
```
default color should be black.

rkusa replied:
Oh right, the first `FF` of the previous color was the alpha part. Will be updated

cmyr replied:
yea I made this mistake too when I was first playing around with styles for the plugin lib. 😎

cmyr commented on XiEditor/StyleMap.swift:
```diff
@@ -61,15 +57,46 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
+        // from CSS-like values that are between 100-900
+        let weight = (json["weight"] as? Int ?? fm.weight(of: fromFont)) / 100 + 1
+
+        if italic {
```
bikeshed question: if we have a true italic and a true bold, how do we prefer to render a bold italic? 🤔

rkusa replied:
This case is handled. Since the font in this of statement takes the weight into account. The second closest match is only called ich this here did not found any match. :)

raphlinus replied:
I think the `fm.weight(of:)` method requires a slightly different conversion. https://bugs.webkit.org/show_bug.cgi?id=6484 has the info but I'll give the table below:

    100 -> 2
    200 -> 3
    300 -> 4
    400 -> 5
    500 -> 6
    600 -> 8
    700 -> 9
    800 -> 10
    900 -> 12

I heard somewhere (not sure where) that 1000 is or will be a valid CSS value. I think it should map to 15.

rkusa replied:
Raph, will be updated, thanks for investigating. @cmyr I think I understood your question wrong. Currently, if there is an italic font, it will be chosen in the best matching weight, regardless of whether there is a better matching weight as non-italic font. We could also choose the better matching weight and apply the Obliqueness attribute instead. I am leaning toward using the true italic font over the better matching weight

cmyr replied:
yea, that was my question. I prefer the current approach of using a true italic if available.

raphlinus commented on XiEditor/StyleMap.swift:
```diff
@@ -108,4 +122,25 @@ class StyleMap {
         }
     }
+
+    func stringWidth(str: NSString, style id: Int) -> CGFloat? {
```
This is intended for consumption by line wrapping?

rkusa replied:
I had this in mind at first, however, using the `.size().width` method of an `NSAttributedString` may be more appropriate to also handle the case where parts of a word have different fonts.

However, it could be useful to fix #21.

But it could probably be be removed from this PR and re-add it once it is actually used.

rkusa commented on XiEditor/StyleMap.swift:
```diff
@@ -61,15 +57,46 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
         let italic = json["italic"] as? Bool ?? false
+        // convert to NSFont weight: (100-500 -> 2-6 (5 normal weight), 600-800 -> 8-10, 900 -> 12
+        let w = Float(json["weight"] as? Int ?? fm.weight(of: fromFont))
+        let weight = Int(floor(1 + w * (0.01 + 3e-6 * w)))
```
@raphlinus Updated to use the suggested formula. Hower, it currently maps `1000` to `14` (and not `15`).

If this is a issue, I could also update the logic to a select, e.g.

```swift
func fontWeight(_ input: Int) -> Int {
    switch input {
    case 100...500:
        return input / 100 + 1
    case 600...800:
        return input / 100 + 2
    case 900:
        return 12
    case 1000:
        return 15
    default:
        return 5
    }
}
```

cmyr replied:
if you only need an exception for that case you could also just do `let weight = w >= 1000 ? 15 : Int(floor(1 + w * (0.01 + 3e-6 * w)))`

rkusa replied:
Yes, that would be less code and I guess the comment above is enough to not require a reader to quickly understand the formula. One (not clearly mentioned) part of my question was, if we want to add this case at all, since it is somehow unclear whether `1000` is already a valid CSS font weight. :)

cmyr replied:
I like having the code self-document but you probably need a comment for the rationale anyway. We aren't _actually_ using css weights, so we can use 1000 if we want to, and it makes sense to have the full system-expressible range available?

I do appreciate that in general any approach I suggest is something you've already tried and decided against for good reasons, but I've got to keep suggesting them. 😉

rkusa replied:
Yes, please keep giving suggestions! 😊

rkusa commented:
Thanks for the reviews! I've pushed a commit addressing the comments.

The remaining open questions is:

Should all usages of `TextDrawingMetrics` in `EditViewController` use a style instead? (requires some more reserved styles like `0 - default`, `1 - selection`, `2- gutter`).
If it helps, I could prepare a PR showing exactly what I have in mind.

raphlinus commented:
I think 14 is actually reasonable, after all 100 maps to 2, not 1.

My gut feeling is that all usages should use a style, and I can see that slightly expanding the pre-defined styles.

rkusa commented on XiEditor/EditView.swift:
```diff
@@ -228,7 +183,7 @@ class EditView: NSView, NSTextInputClient {
             // but that means drawing the selection highlight ourselves (which has other benefits).
             //attrString.drawAtPoint(NSPoint(x: x0, y: y - 13))
-            let y = dataSource.textMetrics.linespace * CGFloat(lineIx + 1);
+            let y = defaultStyle.linespace * CGFloat(lineIx + 1);
```
I am not sure, but would consider it possible that different font weights have a different `linespace`, which is not taken into account here. I hope we can ignore it and just keep using the default style `linespace` 🤔

cmyr replied:
This sounds like a question @raphlinus is very equipped to answer. 😁

cmyr commented on XiEditor/StyleMap.swift:
```diff
@@ -14,13 +14,87 @@
+    var font: NSFont
+    var fgColor: NSColor
+    var bgColor: NSColor
+    var setTextBgColor: Bool
```
`wantsTextBgColor`? 

But I'm not sure I totally understand: text background color should be a distinct property from global background color. The concept of a global background color should exist outside of `Style`, which should more accurately be thought of as `TextStyle` (This renaming might be helpful?). How we end up doing non-text UI styling is something to discuss.

rkusa replied:
Maybe my thinking is somehow influenced, because I've created a Sublime color scheme recently (which is not to be confused with a sublime ui theme). There you have a default foreground and background color, but setting the default text background color gives you

<img width="645" alt="screen shot 2017-05-09 at 07 12 22" src="https://cloud.githubusercontent.com/assets/409021/25836635/c7a41d76-3488-11e7-801f-d2ab01595734.png">

and not

<img width="645" alt="screen shot 2017-05-09 at 07 01 51" src="https://cloud.githubusercontent.com/assets/409021/25836638/cb18a274-3488-11e7-992d-a8758a9fb315.png">


The same applies to the gutter foreground and background color (which are part of the color scheme in Sublime).

<img width="645" alt="screen shot 2017-05-09 at 07 15 15" src="https://cloud.githubusercontent.com/assets/409021/25836640/cf9f8402-3488-11e7-9dd4-aae8bbd650de.png">

vs

<img width="645" alt="screen shot 2017-05-09 at 07 15 24" src="https://cloud.githubusercontent.com/assets/409021/25836643/d27ed7fe-3488-11e7-9e95-79b43251402e.png">

Which I think is the expected behaviour when creating a color scheme and I wouldn't consider both examples a part of a distinct UI-theming.

This example is also where the `setTextBgColor` comes from. Because when I am setting the "global" textview/gutter background, it is not necessary to also set the text background (even though you wouldn't notice because they are of the same color - but it felt wrong).

Again, I am open to the change these things, just want to give my reasoning first.

cmyr commented on XiEditor/EditView.swift:
```diff
@@ -166,9 +114,15 @@ class EditView: NSView, NSTextInputClient {
+
+        // set view background
+        self.wantsLayer = defaultStyle.bgColor.alphaComponent != 0.0
+        self.layer?.backgroundColor = defaultStyle.bgColor.cgColor
```
In any case, I don't think we should be modifying our layer directly here:

[wantsLayer](https://developer.apple.com/reference/appkit/nsview/1483695-wantslayer#):
>In a layer-backed view, any drawing done by the view is cached to the underlying layer object. This cached content can then be manipulated in ways that are more performant than redrawing the view contents explicitly. AppKit automatically creates the underlying layer object (using the makeBackingLayer() method) and handles the caching of the view’s content. If the wantsUpdateLayer method returns false, you should not interact with the underlying layer object directly. Instead, use the methods of this class to make any changes to the view and its layer. If wantsUpdateLayer returns true, it is acceptable (and appropriate) to modify the layer in the view’s updateLayer() method.

At some point in the future I expect us to go down to the CALayer level for all of our drawing, but for now I think we should stick with the higher level drawing API, and just use `NSRectFill`.

Keep the bg stuff in here, I think; when the style changes fire `editView.needsDisplay = True` to invalidate the view's contents, and then it will use the new defaultStyle on the next drawing cycle.

cmyr replied:
Also see my comment about `setTextBgColor`: the `bgColor` field of a style should not be interpreted as the view's background color, as fun as having that working is. 😢

rkusa replied:
`editView.needsDisplay` sounds good. I'll have a look how to set the background with `NSRectFill` instead of using the layer.

cmyr replied:
yes, all drawing has to stay in the draw call.

rkusa commented:
I've updated the implementation. I am not using `Style`s for editor and gutter theming anymore. However, I also haven't created a `ColorTheme` struct as discussed in IRC yesterday. Since there is no coupling between `Style` and `ColorTheme` anymore, since the `ColorTheme` wouldn't be customisable yet (i.e. no extra functionality), and since it would therefore just bloat the PR, I decided to leave it out for now (but I have it in the stash for someday in the future). I also kept the `TextDrawingMetrics` for now (I postponed my thoughts of how to refactor them, mostly to reduce the amount of changes and also since I've not included the `ColorTheme` changes).

I've also already squashed the commits (sorry for that, but having the last changes as a single commit wouldn't help review, because it reverted a lot of single lines).

Ready again for review.

cmyr commented on XiEditor/StyleMap.swift:
```diff
@@ -16,11 +16,55 @@ import Cocoa
+        }
+
+        if font == nil && weight != nil {
+            font = closestMatch(of: fromFont, traits: fm.traits(of: fromFont), weight: weight!) ?? fromFont
```
closest match could just return the input value itself if there's no match. Not sure if I like this better, just musing.

rkusa replied:
Actually, the `?? fromFont` wasn't supposed to be there anymore, since I prefer to not set the `NSFontAttributeName` if the default font is used. I'll remove it

cmyr commented on XiEditor/StyleMap.swift:
```diff
@@ -59,17 +103,28 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
         let italic = json["italic"] as? Bool ?? false
+        var weight = json["weight"] as? Int
+        if let w = weight {
+            // convert to NSFont weight: (100-500 -> 2-6 (5 normal weight), 600-800 -> 8-10, 900 -> 12
```
i'd be tempted to include a link to the discussion issue inline, for future archaeologists

rkusa replied:
you mean this one https://bugs.webkit.org/show_bug.cgi?id=6484 ?

cmyr replied:
I was thinking of the comment from raph explaining the weighting and offered the little formula actually.

cmyr commented on XiEditor/StyleMap.swift:
```diff
@@ -108,4 +150,26 @@ class StyleMap {
+
+    func updateFont(to font: NSFont) {
+        self.font = font
+        map = map.map { $0.map {
```
this line suggests strongly to me that we should rename `map`.

cmyr commented on XiEditor/AppDelegate.swift:
```diff
@@ -14,12 +14,16 @@
 import Cocoa
 
+//TODO: preferred font should be a user preference
+let defaultFont = CTFontCreateWithName("InconsolataGo" as CFString?, 14, nil)
```
Two options:

1. if this is going to be a global, go `FULL_GLOBAL_STYLE`. Otherwise I think it could live comfortably in the app delegate itself, or as a `static` (class) property on `TextDrawingMetrics` (or somewhere else for that matter).

cmyr commented on XiEditor/EditView.swift:
```diff
@@ -42,19 +42,13 @@ struct TextDrawingMetrics {
-        descent = CTFontGetDescent(font)
-        leading = CTFontGetLeading(font)
+        ascent = font.ascender
+        descent = abs(font.descender)
```
Any reason for choosing `abs` here? I suspect there are situations where this could cause `linespace` to be incorrect.

rkusa commented:
@cmyr Thanks for the review! I've added a commit addressing your comments. If a change is not what you may have anticipated, feel free to ask for a change again! 😊

cmyr commented on XiEditor/AppDelegate.swift:
```diff
@@ -14,14 +14,14 @@
     var dispatcher: Dispatcher?
 
+    //TODO: preferred font should be a user preference
+    static let defaultFont = CTFontCreateWithName("InconsolataGo" as CFString?, 14, nil)
```
heh if we're going to put this in app delegate then it should be an instance variable, because app delegate is already a singleton. Sorry if I was unclear 😂

cmyr commented on XiEditor/AppDelegate.swift:
```diff
@@ -90,6 +94,20 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+        for doc in NSApplication.shared().orderedDocuments {
+            guard let doc = doc as? Document else { continue }
+            doc.editViewController?.updateGutterWidth()
+            doc.editViewController?.editView.needsDisplay = true
```
oh, one last thing: we might want, on font change, to re send the `scroll` notification? Otherwise there may be visual jank when changing font sizes.

cmyr replied:
yea so if you change from a font that fits the document in the current window to a font which needs scroll, scrolling isn't available until you next send an update (like by moving the cursor)

cmyr commented:
Okay, looks good to me!

raphlinus commented:
Thanks!
